File: 110 Balanced Binary Tree/balanced_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {	// Author: Xinyu Chen
public:
    bool isBalanced(TreeNode* root) {
        if (root == nullptr) return true;
        if (depth(root) == -1) return false;
        return true;
    }
    
    int depth(TreeNode* node) {
        if (node == nullptr) return 0;
        int depthLeft = depth(node->left);
        if (depthLeft == -1) return -1;
        int depthRight = depth(node->right);
        if (depthRight == -1) return -1;
        int diff = depthLeft - depthRight;
        if (diff > 1 || diff < -1) return -1;
        return max(depthLeft, depthRight) + 1;
    }
};

File: 110 Balanced Binary Tree/balanced_binary_tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {	// Author: Xinyu Chen
    public bool IsBalanced(TreeNode root) {
        if (root == null) return true;
        return dfs(root) != -1;
    }
    
    public int dfs(TreeNode node) {
        if (node == null) return 0;
        int left = dfs(node.left);
        if (left == -1) return -1;
        int right = dfs(node.right);
        if (right == -1) return -1;
        int diff = left - right;
        if (diff < -1 || diff > 1) return -1;
        return Math.Max(left, right) + 1;
    }
}